apis/python/src/tiledb/vector_search/object_readers/directory_reader.py

```diff
@@ -8,6 +18,223 @@
+        if file_path.suffix not in suffixes:
+            return False
+        else:
+            return True
+    return True
+
+
+def find_uris_vfs(
+    search_uri: str,
+    config: Optional[Mapping[str, Any]] = None,
+    include: str = "*",
+    exclude: Sequence[str] = ["[.]*", "*/[.]*"],
+    suffixes: Optional[Sequence[str]] = None,
+    max_files: Optional[int] = None,
+) -> Sequence[str]:
+    with tiledb.scope_ctx(config):
+        vfs = tiledb.VFS(config=config, ctx=tiledb.Ctx(config))
+
+        def find(
+            uri: str,
+            *,
+            include: str = "*",
+            exclude: Sequence[str] = (),
+            suffixes: Optional[Sequence[str]] = None,
+            max_files: Optional[int] = None,
+        ):
+            # Stop if we have found max_count files
+            if max_files is not None and len(results) >= max_files:
+                return
+            listing = vfs.ls(uri)
+            for file_uri in listing:
+                # Avoid infinite recursion
+                if file_uri == uri:
+                    continue
+
+                if vfs.is_dir(file_uri):
+                    find(
+                        file_uri,
+                        include=include,
+                        exclude=exclude,
+                        suffixes=suffixes,
+                        max_files=max_files,
+                    )
+
+                else:
+                    match = match_uri(
+                        search_uri=search_uri,
+                        file_uri=file_uri,
+                        include=include,
+                        exclude=exclude,
+                        suffixes=suffixes,
+                    )
+                    if match:
+                        results.append(file_uri)
+
+        # Add one trailing slash to search_uri
+        search_uri = search_uri.rstrip("/") + "/"
+        results = []
+        find(
+            search_uri,
+            include=include,
+            exclude=exclude,
+            suffixes=suffixes,
+            max_files=max_files,
+        )
+        if max_files is not None:
+            return results[:max_files]
+        else:
+            return results
+
+
+def find_uris_tiledb_group(
+    search_uri: str,
+    config: Optional[Mapping[str, Any]] = None,
+    include: str = "*",
+    exclude: Sequence[str] = ["[.]*", "*/[.]*"],
+    suffixes: Optional[Sequence[str]] = None,
+    max_files: Optional[int] = None,
+    verbose: bool = False,
+) -> Sequence[str]:
+    with tiledb.scope_ctx(config):
+
+        def find(
+            uri: str,
+            path: str,
+            include: str = "*",
+            exclude: Sequence[str] = (),
+            suffixes: Optional[Sequence[str]] = None,
+            max_files: Optional[int] = None,
+        ):
+            # Stop if we have found max_count files
+            if max_files is not None and len(results) >= max_files:
+                return
+            obj_type = tiledb.object_type(uri)
```

Comment:
This needs a scoped context. Likely taking a context as a function parameter would work to pass it as a scoped ctx.

Reply:
Removed all config and context passing from `DirectoryReaders`. 

Readers shouldn't own or manage the TileDB configuration. Rather the config is passed through the ingestion or index functions. Some reason for this:
- Reader arguments are serialized and stored in group metadata. Serializing TileDB config would present a risk of exposing user secrets.
- Multiple users can use the same index with different configs (i.e. each user having their credentials for s3). The Reader and its arguments are the same for all users of an Index

How it works currently:
- Config is passed to `ObjectIndex` methods (`__init__`, `create`, `update_index`).
- We perform all reader operations within `with tiledb.scope_ctx(ctx_or_config=config)`
- Readers should expect a `default_ctx` to be set appropriately

Reply:
Thanks for the explanation!